Clean Specific Rooms Behaviour

boomseklecki opened the pull request:
Hi, this is my first pull request. I've been using your homebridge-roomba2 package for several years and created a custom index.js file for the earlier version but when v1.3 came out I wanted to properly implement cleanRoom with arguments so others could use it. I've added a cleanBehaviour setting to choose to clean everywhere or just specific rooms, config settings for the map and region values necessary to provide an argument to cleanRoom and a javascript file to log the lastCommand properties to console so users can enter them in the settings ui. I was able to build and test the changes and I created a changeset in pnpm but I don't see it on github. I'd appreciate any feedback you might have. Thanks for your consideration.

arminf82 commented:
Hi, is there a way to install the Plugin and including your Pull Request on Homebridge? Would love to have this feature.

karlvr commented on src/accessory.ts:
```diff
@@ -130,8 +132,9 @@ export default class RoombaAccessory implements AccessoryPlugin {
+        this.cleanBehaviour = config.cleanBehaviour !== undefined ? config.cleanBehaviour : "everywhere";
+    	this.mission = config.mission;
         this.stopBehaviour = config.stopBehaviour !== undefined ? config.stopBehaviour : "home";
-
```
Let's retain this blank line

karlvr commented on src/dorita980.d.ts:
```diff
@@ -40,6 +41,7 @@ declare module "dorita980" {
         public on(event: string, callback: () => void): void;
         public off(event: string, callback: unknown): void;
         public clean(): Promise<CommandResult>
+	public cleanRoom(mission: any): Promise<CommandResult>
```
Is it possible to defined an interface type for `Mission`?

karlvr reviewed (changes requested):
Thank you very much for your efforts on this. I'd love to integrate it. Generally it looks nearly ready…

Take a look at the changes as shown in GitHub. It's more obvious here that quite a bit of the indentation in your additions doesn't match the rest of the project. Please correct tabs to spaces and ensure that the indentation is right for all of your changes. I've made a couple of other related suggestions.

Then I'll test it out too and maybe do a beta release so others can test.  

We also need a section for the README to introduce the feature and to document how to set it up using that new command. If you can add at least an outline for that I can tidy it up if needs be!

boomseklecki commented:
Thanks for the feedback. I believe I fixed all the indentation issues, sorry my editor wasn't showing them as different. I like the idea of doing a beta release since I only have two Roombas to test on I might not have considered something that affects other robots.
I discovered two new parameters being returned by my getlastcommand script while I was testing, 'noAutoPasses' and 'twoPasses' both optional boolean parameters that can be passed for each cleaning region so I added them to the config.schema. I also created an interface for the 'mission' object as you suggested and added some documentation to the readme file.
When I was merging your latest updates with my repository I inadvertently created a new branch. It looks like it linked this pull request to the new branch but I haven't figured out how to merge it back to main while preserving the pull request and I think changesets does not like that its a branch (the error I'm getting says 'failed to find where HEAD diverged from main') so I have not been able to create a changeset. Just wanted to call it out in case its an issue.

j000shDotCom commented:
Hey there! I'm also interested in this feature. Are there any other changes requested, @karlvr? I am happy to contribute as well if needed.

kangolo commented:
@donavanbecker I can test this out, I have an i3+ that has room mapping (with a single map).

@boomseklecki .DS_Store shouldn't be included in the PR, maybe add that to .gitignore.

boomseklecki commented:
.DS_Store was in my .gitignore so not sure when it was added. The file has been deleted

karlvr commented:
@boomseklecki thank you very much for this, I have merged it in and committed a few minor changes on top. It's available to test now in v1.5.0-beta.4. Hopefully I didn't break anything in my tinkering! Thanks again.

karlvr commented:
@kindofausername @kangolo @j000shDotCom there's a beta version available to test now if you're still keen, see https://github.com/homebridge-plugins/homebridge-roomba2/wiki/Pre-release-versions for instructions on installing pre-release versions.

kangolo commented:
Sorry for the long delay, I set this up today and it all worked, cleaning 5 rooms.

I just used the official app to start a cleaning job with all the rooms I wanted to, then from the homebridge terminal (using the robot details from the plugin's config):
```bash
cd node_modules/homebridge-roomba2
npm run getlastcommand <robot_blid> <robot_pwd> <robot_ip_address>
```
And used the results to fill in the new values on the config page.

Thanks @karlvr & @boomseklecki
